symbolic-ppdb/src/format/sourcelinks.rs

```diff
@@ -0,0 +1,186 @@
+use std::cmp::Ordering;
+
+use crate::{FormatError, FormatErrorKind};
+
+/// See [Source Link PPDB docs](https://github.com/dotnet/designs/blob/main/accepted/2020/diagnostics/source-link.md#source-link-json-schema).
+#[derive(Clone)]
+pub(crate) struct SourceLinkMappings {
+    rules: Vec<Rule>,
+}
+
+#[derive(Clone)]
+struct Rule {
+    pattern: Pattern,
+    url: String,
+}
+
+#[derive(Clone)]
+enum Pattern {
+    Exact(String),
+    Prefix(String),
+}
+
+impl SourceLinkMappings {
+    pub fn empty() -> Self {
+        SourceLinkMappings { rules: Vec::new() }
+    }
+
+    pub fn new<'a, I>(jsons: I) -> Result<Self, FormatError>
+    where
+        I: Iterator<Item = &'a str>,
+    {
+        let mut result = Self { rules: Vec::new() };
+        for json in jsons {
+            result.add_mappings(json)?;
+        }
+        result.sort();
+        Ok(result)
+    }
+
+    fn add_mappings(&mut self, json: &str) -> Result<(), FormatError> {
+        use serde_json::*;
+        let json: Value = serde_json::from_str(json)
+            .map_err(|e| FormatError::new(FormatErrorKind::InvalidSourceLinkJson, e))?;
+
+        let docs = json
+            .get("documents")
+            .and_then(|v| v.as_object())
+            .ok_or_else(Self::err)?;
+
+        self.rules.reserve(docs.len());
+        for doc in docs.iter() {
```

Comment:
```suggestion
        for (k, v) in docs.iter() {
```

IMO thats easier to read.

Reply:
changed